source/fundamentals/typescript.txt

```diff
@@ -0,0 +1,97 @@
+==========
+TypeScript
+==========
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecols
+
+Overview
+--------
+
+In this guide, you can learn about the **TypeScript** features and limitiations
+of the MongoDB Node.js driver. TypeScript is a strongly typed programming
+language that compiles to JavaScript.
+
+All TypeScript features of the driver are optional. All valid JavaScript
+code written with the driver is also valid TypeScript code. 
+
+For more information, see the
+`TypeScript website <https://www.typescriptlang.org/>`__.
+
+Features
+--------
+
+If you use TypeScript, you can specify a type for some classes in the driver.
+All classes that accept a type paramater in the driver have the default type
+:node-api-4.0:`Document <interfaces/document.html>`.
+
+The following classes accept any type that extends the ``Document``
+interface:
+
+- :node-api-4.0:`Collection <classes/collection.html>`
+- :node-api-4.0:`ChangeStream <classes/changestream.html>`
+
+You can pass a type parameter that extends the ``Document`` interface like this:
+
+.. literalinclude:: /code-snippets/typescript/extend-document.ts
+   :language: typescript
+   :linenos:
+
+The following classes accept any type parameter:
+
+- :node-api-4.0:`FindCursor <classes/findcursor.html>`
+- :node-api-4.0:`AggregationCursor <classes/aggregationcursor.html>`
+
+You can find a code snippet that shows how to specify a type for the ``FindCursor``
+class in the
+:ref:`Find a Document Usage Example <node-driver-findone-usage-example-code-snippet>`.
+
+Limitations
+-----------
+
+.. _node-driver-typescript-limitations-dot-notation:
+
+The driver cannot infer the type of values with keys containing **dot
+notation**. Dot notation is a property access syntax for navigating BSON objects.
+Click on the tabs to see code snippets that highlight this behavior:
```

Comment:
Add here something along the lines of:
This has an additional effect of any keys that are not explicitly listed in your Schema are permitted and typed as any.

i.e.

```typescript
const c = db.collection<{ someKey: number }>('...')
c.find({ someRANDOMKey: 'whatever type I want!' })
```

Reply:
Added an "Important" note describing this behavior!

Is this intended behavior or a limitation? I thought this was intended behavior modeling MongoDBs acceptance of any key in documents?

I added an "Important" note to the "Features" section clarifying this behavior. If you believe this should be described as a "Limitation" (something that in a perfect world should be there but due to a technical or resource constraint is not there at this time), let me know and I'll move the note!

Reply:
Interesting point. In a perfect world where TS is tailored to representing MongoDB query syntax I suppose I don't think we would permit any key in a `Filter<Schema>` because if you want to be able to query a document with a key that isn't in `Schema` you should add it to `Schema` or redeclare the collection you are querying with a new `Schema2`. Also in a perfect world, I also think this would be configurable, user's would opt-in to strict `Filter<X>` vs something more flexible knowing the risks.

But focusing on the docs at hand, I would say this is working as designed then since we do want the flexibility of supporting any key in a query, it wasn't exactly planned that way 😅 the "any key" functionality is specifically to support dot and array notation, but it does work out to be a Feature.

Reply:
Woah thanks for the detailed backstory! Super interesting design decisions. Would be super cool to see opt in behavior for schema typing in the future.